client/trino-client/src/main/java/io/trino/client/spooling/encoding/JsonQueryDataAccess.java

```diff
@@ -13,36 +13,124 @@
 
 public class JsonQueryDataAccess
         implements QueryDataAccess
 {
-    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
-    private static final TypeReference<List<List<Object>>> TYPE = new TypeReference<List<List<Object>>>() {};
-
-    private final List<Column> columns;
     private final InputStream stream;
+    private final TypeDecoder[] decoders;
 
     public JsonQueryDataAccess(List<Column> columns, InputStream stream)
     {
-        this.columns = requireNonNull(columns, "columns is null");
+        this.decoders = createTypeDecoders(columns);
         this.stream = requireNonNull(stream, "stream is null");
     }
 
+    @SuppressModernizer
     @Override
     public Iterable<List<Object>> toIterable()
             throws IOException
     {
-        return fixData(columns, OBJECT_MAPPER.readValue(stream, TYPE));
+        return new RowWiseIterator(stream, decoders);
+    }
+
+    private static class RowWiseIterator
+            implements Iterable<List<Object>>, Iterator<List<Object>>
+    {
+        private final InputStream stream;
+        private boolean closed;
+        private final JsonParser parser;
+        private final TypeDecoder[] decoders;
+
+        @SuppressModernizer // There is no JsonFactory in the client module
+        public RowWiseIterator(InputStream stream, TypeDecoder[] decoders)
+                throws IOException
+        {
+            this.stream = requireNonNull(stream, "stream is null");
+            this.parser = new JsonFactory()
+                    .enable(JsonParser.Feature.USE_FAST_DOUBLE_PARSER)
+                    .enable(JsonParser.Feature.USE_FAST_BIG_NUMBER_PARSER)
+                    .disable(JsonParser.Feature.AUTO_CLOSE_SOURCE) // We want to close source explicitly
+                    .createParser(stream);
+            this.decoders = requireNonNull(decoders, "decoders is null");
+
+            try {
+                verify(parser.nextToken() == START_ARRAY, "Expected start of array, but got %s", parser.currentToken());
+                verify(parser.nextToken() == START_ARRAY, "Expected start of array, but got %s", parser.currentToken());
+            }
+            catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        private void advanceEndValues()
+        {
+            try {
+                verify(parser.nextToken() == END_ARRAY, "Expected end of array, but got %s", parser.currentToken());
+                switch (parser.nextToken()) {
+                    case END_ARRAY:
+                        closed = true;
+                        stream.close();
+                        break;
+                    case START_ARRAY:
+                        closed = false;
+                        break;
+                    default:
+                        throw new IllegalStateException("Expected end of array or start of array, but got " + parser.currentToken());
```

Comment:
Throwing `JsonProcessingException` seems more appropriate

Or throw `VerifyException` as that matches the `verify()` above